# Venu_SticklebackRecombination

Fine-scale contemporary recombination variation and its fitness consequences in adaptively diverging stickleback fish

Vrinda Venu1,*, Enni Harjunmaa1, Andreea Dreau1, Shannon Brady2, Devin Absher3, David M. Kingsley2,4, Felicity C. Jones1,5,†

Affiliations
1.	Friedrich Miescher Laboratory of the Max Planck Society, Max Planck Ring 9, 72076 Tübingen, Germany
2.	Dept. of Developmental Biology, Stanford University School of Medicine, Stanford, CA 94305-5329, USA
3.	HudsonAlpha Institute for Biotechnology, 601 Genome Way, Huntsville, AL 35806, USA
4.	Howard Hughes Medical Institute, Chevy Chase, MD 20815, USA
5.	Groningen Institute for Evolutionary Life Sciences (GELIFES), University of Groningen, Nijenborgh 7, Groningen, The Netherlands

\* vrinda@lanl.gov

† corresponding author: fcjones@tuebingen.mpg.de

Abstract

Despite deep evolutionary conservation, recombination varies greatly across the genome, among individuals, sexes and populations and can be a major evolutionary force in the wild. Yet this variation in recombination and its impact on adaptively diverging populations is not well understood. To elucidate the nature and potential consequences of recombination rate variation, we characterized fine-scale recombination landscapes by combining pedigrees, functional genomics and field fitness measurements in an adaptively divergent pair of marine and freshwater threespine stickleback populations from River Tyne, Scotland. Through whole-genome sequencing of large nuclear families, we identified the genomic location of almost 50,000 crossovers and built recombination maps for 36 marine, freshwater, and hybrid individuals at 3.8 kilobase resolution. Using these maps, we quantified the factors driving variation in recombination rate: we find strong heterochiasmy between sexes (68% of the variation) but also differences among ecotypes (21.8%). Hybrids show evidence of significant recombination suppression, both in overall map length and in individual loci. We further tested and found reduced recombination rates both within single marine–freshwater adaptive loci and between loci on the same chromosome, suggestive of selection on linked ‘cassettes’. We tested theory supporting the evolution of linked selection using temporal sampling along a natural hybrid zone, and found that recombinants with shuffled alleles across loci show traits associated with reduced fitness. Our results support predictions that divergence in cis-acting recombination modifiers whose mechanisms are disrupted in hybrids, may have an important role to play in the maintenance of differences among adaptively diverging populations.


Code

This Github provides the code for population genetic simulations exploring the effects of recombination heterochiasmy and recombination suppression in cis(coupled) heterozygotes in the context of populations undergoing a adaptive divergence with gene flow with a selection-migration balance.  

The relevant Rscript can be found here: Recombination.SelMigBalance.Rscript

Further details and description are outlined in the supplementary methods of the paper.
